src/box/lua/trigger.c

```diff
@@ -0,0 +1,242 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#include "box/lua/func_adapter.h"
+#include "core/event.h"
+#include <diag.h>
+#include <fiber.h>
+#include "lua/utils.h"
+
+/**
+ * Sets a trigger with passed name to the passed event.
+ * The first argument is event name, the second one is trigger name, the third
+ * one is new trigger handler - it can be a Lua function or another callable
+ * object. If there is an already registered trigger with such name in the
+ * event, it is replaced with new trigger.
+ * Returns new trigger handler (which was passed as the third argument).
+ */
+static int
+luaT_trigger_set(struct lua_State *L)
+{
+	const char *event_name = luaL_checkstring(L, 1);
+	const char *trigger_name = luaL_checkstring(L, 2);
+	if (lua_gettop(L) < 3 || !luaL_iscallable(L, 3))
+		luaL_typerror(L, 3, "callable");
+	/*
+	 * The following code is written in assumption no error will be thrown.
+	 */
+	struct event *event = event_get(event_name, true);
+	assert(event != NULL);
+	struct func_adapter *func = func_adapter_lua_create(L, 3);
+	event_reset_trigger(event, trigger_name, func);
+	func_adapter_lua_get_func(func, L);
+	return 1;
+}
+
+/**
+ * Deletes a trigger with passed name from passed event.
+ * The first argument is event name, the second one is trigger name.
+ * Returns deleted trigger handler.
+ */
+static int
+luaT_trigger_del(struct lua_State *L)
+{
+	const char *event_name = luaL_checkstring(L, 1);
+	struct event *event = event_get(event_name, false);
+	if (event == NULL)
+		return 0;
+	const char *trigger_name = luaL_checkstring(L, 2);
+	struct func_adapter *old = event_find_trigger(event, trigger_name);
+	if (old != NULL)
+		func_adapter_lua_get_func(old, L);
+	else
+		lua_pushnil(L);
+	event_reset_trigger(event, trigger_name, NULL);
+	return 1;
+}
+
+/**
+ * Calls all the triggers registered on passed event with variable number of
+ * arguments. Execution is stopped by a first exception.
+ * First argument must be a string - all the other arguments will be passed
+ * to the triggers without any processing or copying.
+ * Returns no values on success. If one of the triggers threw an error, it is
+ * raised again.
+ */
+static int
+luaT_trigger_call(struct lua_State *L)
+{
+	const char *event_name = luaL_checkstring(L, 1);
+	struct event *event = event_get(event_name, false);
+	if (event == NULL)
+		return 0;
+	int top = lua_gettop(L);
+	int narg = top - 1;
+	struct event_trigger_iterator it;
+	event_trigger_iterator_create(&it, event);
+	struct func_adapter *trg = NULL;
+	const char *name = NULL;
+	int rc = 0;
+	while (rc == 0 && event_trigger_iterator_next(&it, &trg, &name)) {
+		lua_settop(L, top);
+		func_adapter_lua_get_func(trg, L);
+		for (int i = top - narg + 1; i <= top; ++i)
+			lua_pushvalue(L, i);
+		rc = luaT_call(L, narg, LUA_MULTRET);
```

Comment:
You don't use the result so you could just set the `nresults` argument to 0 instead of `LUA_MULTRET`. Then you wouldn't need to call `lua_settop` at each iteration.

Reply:
Done.